### PR TITLE
CLOUDFRONT_VIEWER_POLICY_HTTPS

### PR DESCRIPTION
**Pull request by vmunugal**
I confirm these files are made available under CC0 1.0 Universal (https://creativecommons.org/publicdomain/zero/1.0/legalcode)

*Issue #, if available:*

*Description of changes:*

### CODE REVIEW

File: python/CLOUDFRONT_VIEWER_POLICY_HTTPS/CLOUDFRONT_VIEWER_POLICY_HTTPS.py [hunk 1]
```diff
@@ -0,0 +1,367 @@
...
+CONFIG_ROLE_TIMEOUT_SECONDS = 900
+
+#############
+# Main Code #
+#############
+
+def evaluate_compliance(event, configuration_item, valid_rule_parameters):
+    distribution_id = configuration_item['resourceId']
+    if configuration_item['configuration']['distributionConfig']['defaultCacheBehavior']['viewerProtocolPolicy'] == 'allow-all':
+        return build_evaluation(distribution_id, 'NON_COMPLIANT', event, DEFAULT_RESOURCE_TYPE, annotation='''Default ViewerProtocolPolicy is set to 'allow-all' for this distribution.''')
```
Comment (shikharj05) [inline python/CLOUDFRONT_VIEWER_POLICY_HTTPS/CLOUDFRONT_VIEWER_POLICY_HTTPS.py]: Change annotation to full service name: Default ViewerProtocolPolicy is set to 'allow-all' for this Amazon CloudFront distribution.

File: python/CLOUDFRONT_VIEWER_POLICY_HTTPS/CLOUDFRONT_VIEWER_POLICY_HTTPS.py [hunk 2]
```diff
@@ -0,0 +1,367 @@
...
+    if configuration_item['configuration']['distributionConfig']['cacheBehaviors']['quantity'] > 0:
+        for behavior in configuration_item['configuration']['distributionConfig']['cacheBehaviors']['items']:
+            path = behavior['pathPattern']
+            if behavior['viewerProtocolPolicy'] == 'allow-all':
+                return build_evaluation(distribution_id, 'NON_COMPLIANT', event, DEFAULT_RESOURCE_TYPE, annotation='''ViewerProtocolPolicy for path "{}" is set to 'allow-all.' '''.format(path))
```
Comment (shikharj05) [inline python/CLOUDFRONT_VIEWER_POLICY_HTTPS/CLOUDFRONT_VIEWER_POLICY_HTTPS.py]: Change annotation to: ViewerProtocolPolicy for CacheBehavior with path pattern "{}" is set to 'allow-all.' '''.format(path)

File: python/CLOUDFRONT_VIEWER_POLICY_HTTPS/CLOUDFRONT_VIEWER_POLICY_HTTPS_test.py [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+except ImportError:
+    from mock import MagicMock
+import botocore
+
+##############
+# Parameters #
+##############
+
+# Define the default resource to report to Config Rules
+DEFAULT_RESOURCE_TYPE = 'AWS::::Account'
```
Comment (shikharj05) [inline python/CLOUDFRONT_VIEWER_POLICY_HTTPS/CLOUDFRONT_VIEWER_POLICY_HTTPS_test.py]: Update DEFAULT_RESOURCE_TYPE to AWS::CloudFront::Distribution

File: python/CLOUDFRONT_VIEWER_POLICY_HTTPS/CLOUDFRONT_VIEWER_POLICY_HTTPS_test.py [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+
+RULE = __import__('CLOUDFRONT_VIEWER_POLICY_HTTPS')
+
+class SampleTest(unittest.TestCase):
+
+    rule_parameters = '{"SomeParameterKey":"SomeParameterValue","SomeParameterKey2":"SomeParameterValue2"}'
+
+    invoking_event_iam_role_sample = '{"configurationItem":{"relatedEvents":[],"relationships":[],"configuration":{},"tags":{},"configurationItemCaptureTime":"2018-07-02T03:37:52.418Z","awsAccountId":"123456789012","configurationItemStatus":"ResourceDiscovered","resourceType":"AWS::IAM::Role","resourceId":"some-resource-id","resourceName":"some-resource-name","ARN":"some-arn"},"notificationCreationTime":"2018-07-02T23:05:34.445Z","messageType":"ConfigurationItemChangeNotification"}'
+
+    resource_id = ['E10KYRP2JT92H5', 'E09KYRP2JT92H5', 'E08KYRP2JT92H5']
```
Comment (shikharj05) [inline python/CLOUDFRONT_VIEWER_POLICY_HTTPS/CLOUDFRONT_VIEWER_POLICY_HTTPS_test.py]: Change resource IDs to mock values.

File: python/CLOUDFRONT_VIEWER_POLICY_HTTPS/CLOUDFRONT_VIEWER_POLICY_HTTPS_test.py [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+
+    configurations = [{'distributionConfig':{'cacheBehaviors':{'quantity':2, 'items':[{'viewerProtocolPolicy':'allow-all', 'pathPattern':'images/*.jpg'}, {'viewerProtocolPolicy':'allow-all', 'pathPattern':'videos/*.mp4'}]}, 'defaultCacheBehavior': {'viewerProtocolPolicy':'redirect-to-https'}}}, {'distributionConfig':{'cacheBehaviors':{'quantity':0}, 'defaultCacheBehavior':{'viewerProtocolPolicy':'allow-all'}}}, {'distributionConfig':{'cacheBehaviors':{'quantity':0}, 'defaultCacheBehavior':{'viewerProtocolPolicy':'redirect-to-https'}}}]
+    def setUp(self):
```
Comment (shikharj05) [inline python/CLOUDFRONT_VIEWER_POLICY_HTTPS/CLOUDFRONT_VIEWER_POLICY_HTTPS_test.py]: Remove this test case

File: python/CLOUDFRONT_VIEWER_POLICY_HTTPS/CLOUDFRONT_VIEWER_POLICY_HTTPS_test.py [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+        pass
+
+    def test_sample(self):
```
Comment (shikharj05) [inline python/CLOUDFRONT_VIEWER_POLICY_HTTPS/CLOUDFRONT_VIEWER_POLICY_HTTPS_test.py]: remove this one as well

File: python/CLOUDFRONT_VIEWER_POLICY_HTTPS/CLOUDFRONT_VIEWER_POLICY_HTTPS_test.py [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+        self.assertTrue(True)
+
+# Custom cache behavior NON_COMPLIANT
```
Comment (shikharj05) [inline python/CLOUDFRONT_VIEWER_POLICY_HTTPS/CLOUDFRONT_VIEWER_POLICY_HTTPS_test.py]: match the comment indentation with function indentation for all 3 test scenario and label them as Gherkin Scenario 1a, Gherkin Scenario 1b & Gherkin Scenario 2 with short descriptions

File: python/CLOUDFRONT_VIEWER_POLICY_HTTPS/CLOUDFRONT_VIEWER_POLICY_HTTPS_test.py [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+    def test_scenario1(self):
+        invoking_event = construct_invoking_event(construct_config_item(self.configurations[0], self.resource_id[0]))
+        lambda_event = build_lambda_configurationchange_event(invoking_event, rule_parameters=None)
+        response = RULE.lambda_handler(lambda_event, {})
+        resp_expected = []
+        resp_expected.append(build_expected_response('NON_COMPLIANT', 'E10KYRP2JT92H5', 'AWS::CloudFront::Distribution', annotation='''ViewerProtocolPolicy for path "images/*.jpg" is set to 'allow-all.' '''))
+        assert_successful_evaluation(self, response, resp_expected, 1)
```
Comment (shikharj05) [inline python/CLOUDFRONT_VIEWER_POLICY_HTTPS/CLOUDFRONT_VIEWER_POLICY_HTTPS_test.py]: default value for count is 1, no need to mention it.

File: python/CLOUDFRONT_VIEWER_POLICY_HTTPS/CLOUDFRONT_VIEWER_POLICY_HTTPS_test.py [hunk 9]
```diff
@@ -0,0 +1,218 @@
...
+
+# Default cache behavior NON_COMPLIANT
+    def test_scenario2a(self):
+        invoking_event = construct_invoking_event(construct_config_item(self.configurations[1], self.resource_id[1]))
+        lambda_event = build_lambda_configurationchange_event(invoking_event, rule_parameters=None)
+        response = RULE.lambda_handler(lambda_event, {})
+        resp_expected = []
+        resp_expected.append(build_expected_response('NON_COMPLIANT', 'E09KYRP2JT92H5', 'AWS::CloudFront::Distribution', annotation='''Default ViewerProtocolPolicy is set to 'allow-all' for this distribution.'''))
+        assert_successful_evaluation(self, response, resp_expected, 1)
```
Comment (shikharj05) [inline python/CLOUDFRONT_VIEWER_POLICY_HTTPS/CLOUDFRONT_VIEWER_POLICY_HTTPS_test.py]: default value for count is 1, no need to mention it.

File: python/CLOUDFRONT_VIEWER_POLICY_HTTPS/CLOUDFRONT_VIEWER_POLICY_HTTPS_test.py (hunk 4 above)
Comment (vmunugal) [reply python/CLOUDFRONT_VIEWER_POLICY_HTTPS/CLOUDFRONT_VIEWER_POLICY_HTTPS_test.py]: Resource ID's are mocked values

File: python/CLOUDFRONT_VIEWER_POLICY_HTTPS/CLOUDFRONT_VIEWER_POLICY_HTTPS.py [hunk 10]
```diff
@@ -0,0 +1,367 @@
...
+def evaluate_compliance(event, configuration_item, valid_rule_parameters):
+    distribution_id = configuration_item['resourceId']
+    if configuration_item['configuration']['distributionConfig']['defaultCacheBehavior']['viewerProtocolPolicy'] == 'allow-all':
+        return build_evaluation(distribution_id, 'NON_COMPLIANT', event, DEFAULT_RESOURCE_TYPE, annotation='''Default ViewerProtocolPolicy is set to 'allow-all' for this Amazon CloudFront distribution.''')
+    if configuration_item['configuration']['distributionConfig']['cacheBehaviors']['quantity'] > 0:
+        for behavior in configuration_item['configuration']['distributionConfig']['cacheBehaviors']['items']:
+            path = behavior['pathPattern']
+            if behavior['viewerProtocolPolicy'] == 'allow-all':
+                return build_evaluation(distribution_id, 'NON_COMPLIANT', event, DEFAULT_RESOURCE_TYPE, annotation='''ViewerProtocolPolicy for CacheBehavior with path pattern "{}" is set to 'allow-all.' '''.format(path))
+    return build_evaluation(distribution_id, 'COMPLIANT', event)
```
Comment (shikharj05) [inline python/CLOUDFRONT_VIEWER_POLICY_HTTPS/CLOUDFRONT_VIEWER_POLICY_HTTPS.py]: You need to use build_evaluation_from_config_item() as this is a configuration change rule.

File: python/CLOUDFRONT_VIEWER_POLICY_HTTPS/CLOUDFRONT_VIEWER_POLICY_HTTPS.py [hunk 11]
```diff
@@ -0,0 +1,367 @@
...
+CONFIG_ROLE_TIMEOUT_SECONDS = 900
+
+#############
+# Main Code #
+#############
+
+def evaluate_compliance(event, configuration_item, valid_rule_parameters):
+    distribution_id = configuration_item['resourceId']
+    if configuration_item['configuration']['distributionConfig']['defaultCacheBehavior']['viewerProtocolPolicy'] == 'allow-all':
+        return build_evaluation(distribution_id, 'NON_COMPLIANT', event, DEFAULT_RESOURCE_TYPE, annotation='''Default ViewerProtocolPolicy is set to 'allow-all' for this Amazon CloudFront distribution.''')
```
Comment (shikharj05) [inline python/CLOUDFRONT_VIEWER_POLICY_HTTPS/CLOUDFRONT_VIEWER_POLICY_HTTPS.py]: You need to use build_evaluation_from_config_item() as this is a configuration change rule.

**Comment by shikharj05** [comment]
@jongogogo Kindly check the code and provide your feedback, the rule looks good to me!

File: python/CLOUDFRONT_VIEWER_POLICY_HTTPS/CLOUDFRONT_VIEWER_POLICY_HTTPS_test.py [hunk 12]
```diff
@@ -0,0 +1,210 @@
...
+        invoking_event = construct_invoking_event(construct_config_item(self.configurations[1], self.resource_id[1]))
+        lambda_event = build_lambda_configurationchange_event(invoking_event, rule_parameters=None)
+        response = RULE.lambda_handler(lambda_event, {})
+        resp_expected = []
+        resp_expected.append(build_expected_response('NON_COMPLIANT', 'E1234567891234', 'AWS::CloudFront::Distribution', annotation='''Default ViewerProtocolPolicy is set to 'allow-all' for this Amazon CloudFront distribution.'''))
+        assert_successful_evaluation(self, response, resp_expected)
+
+    #Gerkin Scenario 1b: COMPLIANT
+    def test_scenario1b(self):
+        invoking_event = construct_invoking_event(construct_config_item(self.configurations[2], self.resource_id[2]))
```
Comment (shikharj05) [inline python/CLOUDFRONT_VIEWER_POLICY_HTTPS/CLOUDFRONT_VIEWER_POLICY_HTTPS_test.py]: combine both functions into one construct_invoking_event() and move them above "#helper functions" comment.

File: python/CLOUDFRONT_VIEWER_POLICY_HTTPS/CLOUDFRONT_VIEWER_POLICY_HTTPS_test.py [hunk 13]
```diff
@@ -0,0 +1,210 @@
...
+        'relatedEvents': [],
+        'relationships': [],
+        'configuration': config,
+        'configurationItemVersion': None,
+        'configurationItemCaptureTime': "2019-03-17T03:37:52.418Z",
+        'supplementaryConfiguration': {},
+        'configurationStateId': 1532049940079,
+        'awsAccountId': "SAMPLE",
+        'configurationItemStatus': "ResourceDiscovered",
+        'resourceType': "AWS::CloudFront:Distribution",
```
Comment (shikharj05) [inline python/CLOUDFRONT_VIEWER_POLICY_HTTPS/CLOUDFRONT_VIEWER_POLICY_HTTPS_test.py]: should be AWS::CloudFront::Distribution

File: python/CLOUDFRONT_VIEWER_POLICY_HTTPS/CLOUDFRONT_VIEWER_POLICY_HTTPS_test.py [hunk 14]
```diff
@@ -0,0 +1,206 @@
...
+            return CONFIG_CLIENT_MOCK
+        if client_name == 'sts':
+            return STS_CLIENT_MOCK
+        raise Exception("Attempting to create an unknown client")
+
+sys.modules['boto3'] = Boto3Mock()
+
+RULE = __import__('CLOUDFRONT_VIEWER_POLICY_HTTPS')
+
+class SampleTest(unittest.TestCase):
```
Comment (shikharj05) [inline python/CLOUDFRONT_VIEWER_POLICY_HTTPS/CLOUDFRONT_VIEWER_POLICY_HTTPS_test.py]: One last change from my end, change the name to ComplianceTest

Comment (vmunugal) [reply python/CLOUDFRONT_VIEWER_POLICY_HTTPS/CLOUDFRONT_VIEWER_POLICY_HTTPS_test.py]: Done !!

**Comment by shikharj05** [comment]
Thank you for the changes, @jongogogo kindly check this PR & let us know your feedback.

File: python/CLOUDFRONT_VIEWER_POLICY_HTTPS/CLOUDFRONT_VIEWER_POLICY_HTTPS_test.py [hunk 15]
```diff
@@ -0,0 +1,206 @@
...
+            return STS_CLIENT_MOCK
+        raise Exception("Attempting to create an unknown client")
+
+sys.modules['boto3'] = Boto3Mock()
+
+RULE = __import__('CLOUDFRONT_VIEWER_POLICY_HTTPS')
+
+class ComplianceTest(unittest.TestCase):
+
+    invoking_event_iam_role_sample = '{"configurationItem":{"relatedEvents":[],"relationships":[],"configuration":{},"tags":{},"configurationItemCaptureTime":"2018-07-02T03:37:52.418Z","awsAccountId":"123456789012","configurationItemStatus":"ResourceDiscovered","resourceType":"AWS::IAM::Role","resourceId":"some-resource-id","resourceName":"some-resource-name","ARN":"some-arn"},"notificationCreationTime":"2018-07-02T23:05:34.445Z","messageType":"ConfigurationItemChangeNotification"}'
```
Comment (jongogogo) [inline python/CLOUDFRONT_VIEWER_POLICY_HTTPS/CLOUDFRONT_VIEWER_POLICY_HTTPS_test.py]: remove

File: python/CLOUDFRONT_VIEWER_POLICY_HTTPS/CLOUDFRONT_VIEWER_POLICY_HTTPS.py [hunk 16]
```diff
@@ -0,0 +1,366 @@
...
+#
+#        http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for
+# the specific language governing permissions and limitations under the License.
+
+"""
+Description:
+  Checks whether your Amazon CloudFront Distributions use HTTPS (directly or via a redirection). The rule returns NON_COMPLIANT if the ViewerProtocolPolicy is set to 'allow-all' (i.e Viewers can use HTTP or HTTPS).
```
Comment (jongogogo) [inline python/CLOUDFRONT_VIEWER_POLICY_HTTPS/CLOUDFRONT_VIEWER_POLICY_HTTPS.py]: Can you add `in the defaultCacheBehavior or in the cacheBehaviors`.

`Checks whether your Amazon CloudFront Distributions use HTTPS (directly or via a redirection). The rule returns NON_COMPLIANT if the ViewerProtocolPolicy is set to 'allow-all' in the defaultCacheBehavior or in the cacheBehaviors (i.e Viewers can use HTTP or HTTPS).`

File: python/CLOUDFRONT_VIEWER_POLICY_HTTPS/CLOUDFRONT_VIEWER_POLICY_HTTPS.py [hunk 17]
```diff
@@ -0,0 +1,366 @@
...
+Trigger:
+  Configuration Change on AWS::CloudFront::Distribution
+Reports on:
+  AWS::CloudFront::Distribution
+Rule Parameters:
+  None
+Scenarios:
```
Comment (jongogogo) [inline python/CLOUDFRONT_VIEWER_POLICY_HTTPS/CLOUDFRONT_VIEWER_POLICY_HTTPS.py]: Can you split the non-compliant scenario, distinguishing defaultCache behavior and non-default cache behavior

**Comment by shikharj05** [comment]
@vmunugal has made the changes. @jongogogo Kindly review :D

**Pull request merged**